### PR TITLE
doc: modify stacks documentation

### PR DESCRIPTION
**Pull request by neeraj-laad**
First pass at modify stack document.

Signed-off-by: Neeraj Laad <neeraj.laad@gmail.com>

## Checklist

- [x] commit message follows [commit guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).
- [x] website builds and runs in production - for information on how to test locally go [here](https://github.com/appsody/website/blob/master/DEVELOPMENT.md).

## Related Issues
Related to #358

### CODE REVIEW

File: content/docs/stacks/modify.md [hunk 1]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
 ---
 
 # Modifying a Stack
-
 While there are many Appsody stacks to choose from, you may want to alter some aspects of an existing stack to match your development needs or standards. The modifications can impact how users will develop, build and deploy applications using the stack.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Removed future tense and fixed grammar whilst we're in this file...

"While there are many Appsody stacks to choose from, you might want to alter some aspects of an existing stack to match your development needs or standards. The modifications can impact how users develop, build, and deploy applications that use the stack."

File: content/docs/stacks/modify.md [hunk 2]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
 
 Once you have made the necessary modifications, you can [package](/content/docs/stacks/package.md) the modified stack and test it locally using Appsody CLI.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Fixed grammar...

"After you make your modifications, you can [package](/content/docs/stacks/package.md) the modified stack and test it locally using the Appsody CLI."

File: content/docs/stacks/modify.md [hunk 3]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
 
-When you have ensured the stack is working as intended, you can contribute the modifications to the existing Appsody stack or maintain a separate version with your modifications.
+## Project templates
+Templates provide an initial application to enable developers to get started with a stack. Developers can expand and adapt this template to suit their application needs.
+
+The `templates` directory contains one or more starter applications that are given to the user when they initialise their projects. Every template is contained within its directory, `/templates/<template-name>`.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Do we mean "given to the user" or "created for the user"?
initialise -> initialize

File: content/docs/stacks/modify.md [hunk 4]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
 
 # Modifying a Stack
-
 While there are many Appsody stacks to choose from, you may want to alter some aspects of an existing stack to match your development needs or standards. The modifications can impact how users will develop, build and deploy applications using the stack.
 
 Once you have made the necessary modifications, you can [package](/content/docs/stacks/package.md) the modified stack and test it locally using Appsody CLI.
 
-When you have ensured the stack is working as intended, you can contribute the modifications to the existing Appsody stack or maintain a separate version with your modifications.
+## Project templates
+Templates provide an initial application to enable developers to get started with a stack. Developers can expand and adapt this template to suit their application needs.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: -> "the needs of their application" is clearer?

File: content/docs/stacks/modify.md [hunk 5]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
+
+The `templates` directory contains one or more starter applications that are given to the user when they initialise their projects. Every template is contained within its directory, `/templates/<template-name>`.
+
+Capabilities that apply to all templates are better suited for inclusion in the stack image.
+
+## Stack image for local development
+Appsody utilizes a containerized environment during local development. The stack image defines this environment and specifies the stack behaviour during application development lifecycle.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Appsody utilizes -> Appsody uses
behaviour -> behavior

Comment (helenmasters) [reply content/docs/stacks/modify.md]: during application development lifecycle -> during the application development lifecycle

File: content/docs/stacks/modify.md [hunk 6]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
+
+The stack image contains common capabilities that can be leveraged by all templates. For example, the [`nodejs-express`](https://github.com/appsody/stacks/tree/master/incubator/nodejs-express) stack image provides health endpoints and Prometheus metrics without developers needing to implement it themselves.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: leveraged -> used

File: content/docs/stacks/modify.md [hunk 7]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
+
+The `image` directory contains files for the stack image. The `image/Dockerfile-stack` defines the exact steps for building the stack image.
+
+### Handling file mounts
+During local development, the application code is held on the local file system and is mounted in the running container for the stack. Stack creators configure the `APPSODY_MOUNTS` [environment variable](/content/docs/stacks/environment-variables.md) to specify a list of mount paths to achieve this behaviour.
+
+### Stack environment variables
+Stack creators configure [environment variables](/content/docs/stacks/environment-variables.md) in `Dockerfile-stack` to specify the commands for running, debugging and testing the application. `Appsody CLI` and `Appsody controller` inspect these environment variables and then drive the expected behaviour for the developer during local development.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: behaviour -> behavior

File: content/docs/stacks/modify.md (hunk 6 above)
Comment (helenmasters) [inline content/docs/stacks/modify.md]: "...Prometheus metrics so the developers do not need to implement them." is clearer?

File: content/docs/stacks/modify.md [hunk 8]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
+
+## Stack image for local development
+Appsody utilizes a containerized environment during local development. The stack image defines this environment and specifies the stack behaviour during application development lifecycle.
+
+The stack image contains common capabilities that can be leveraged by all templates. For example, the [`nodejs-express`](https://github.com/appsody/stacks/tree/master/incubator/nodejs-express) stack image provides health endpoints and Prometheus metrics without developers needing to implement it themselves.
+
+The `image` directory contains files for the stack image. The `image/Dockerfile-stack` defines the exact steps for building the stack image.
+
+### Handling file mounts
+During local development, the application code is held on the local file system and is mounted in the running container for the stack. Stack creators configure the `APPSODY_MOUNTS` [environment variable](/content/docs/stacks/environment-variables.md) to specify a list of mount paths to achieve this behaviour.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: behaviour -> behavior

Comment (helenmasters) [reply content/docs/stacks/modify.md]: It isn't clear from this description why you would want to add mounts.

File: content/docs/stacks/modify.md [hunk 9]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
+
+### Stack environment variables
+Stack creators configure [environment variables](/content/docs/stacks/environment-variables.md) in `Dockerfile-stack` to specify the commands for running, debugging and testing the application. `Appsody CLI` and `Appsody controller` inspect these environment variables and then drive the expected behaviour for the developer during local development.
+
+### Monitoring file changes
+Stack creators configure [environment variables](/content/docs/stacks/environment-variables.md) in `Dockerfile-stack` to specify which files are monitored for changes and how to reflect those changes in the running application.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: We should specify which env vars these are, otherwise users will have to search through the whole table.

File: content/docs/stacks/modify.md [hunk 10]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
 
+### Managing dependencies
+Appsody provides the ability to cache any installed dependencies across runs to accelerate local development. This is achieved by creating and mounting a volume independent of a specific container instance.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Can we simplify "provides the ability to cache" to "enables the caching of"?
"across runs" isn't clear. 
"This is achieved by creating and mounting a volume independent of a specific container instance." - it is not clear what we mean here?

File: content/docs/stacks/modify.md [hunk 11]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
-## Modifying the stack image
-The `image` directory contains files for building the stack image. This image controls how the stack behaves throughout the application lifecycle.
+Stack creators configure the `APPSODY_DEPS` [environment variable](/content/docs/stacks/environment-variables.md) to specify the directory to be cached.
 
+## Application build and deployment
+Stack creators provide `Dockerfile` for building a container image for Appsody applications which includes capabilities from the stack and developer's application.
 
-## Modifying stack templates
-The `templates` directory contains one or more starter applications that are given to the user when they initialise their projects. 
+`Appsody CLI` uses this `Dockerfile` to create the application container image when [`appsody build`](/content/docs/using-appsody/cli-commands.md/#appsody-build) command is run.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Suggest rewording to "The Appsody CLI uses the `Dockerfile` to create the application container image when the [`appsody build`](/content/docs/using-appsody/cli-commands.md/#appsody-build) command is run."

File: content/docs/stacks/modify.md [hunk 12]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
 
-Whilst modifying templates, consider if the functionality may be better placed in the stack image itself to affect all templates.
+Stack creators also provide a template deployment manifest `image/config/app-deploy.yaml` that `Appsody CLI` uses to support deployments to Kubernetes or Knative platforms using the `Appsody Operator`.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: that `Appsody CLI` -> that the Appsody CLI

Comment (helenmasters) [reply content/docs/stacks/modify.md]: also why is `Appsody Operator` in quotes?

File: content/docs/stacks/modify.md [hunk 13]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
 
+### Managing dependencies
+Appsody provides the ability to cache any installed dependencies across runs to accelerate local development. This is achieved by creating and mounting a volume independent of a specific container instance.
 
-## Modifying the stack image
-The `image` directory contains files for building the stack image. This image controls how the stack behaves throughout the application lifecycle.
+Stack creators configure the `APPSODY_DEPS` [environment variable](/content/docs/stacks/environment-variables.md) to specify the directory to be cached.
 
+## Application build and deployment
+Stack creators provide `Dockerfile` for building a container image for Appsody applications which includes capabilities from the stack and developer's application.
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Needs rewording a little, maybe... "Stack creators must provide a `Dockerfile` that defines how to build the container image for the Appsody application, including both capabilities from the stack, and the developer's application." ?

File: content/docs/stacks/modify.md [hunk 14]
```diff
@@ -3,19 +3,41 @@ title: Modifying a Stack
...
+Stack creators configure [environment variables](/content/docs/stacks/environment-variables.md) in `Dockerfile-stack` to specify which files are monitored for changes and how to reflect those changes in the running application.
 
+### Managing dependencies
+Appsody provides the ability to cache any installed dependencies across runs to accelerate local development. This is achieved by creating and mounting a volume independent of a specific container instance.
 
-## Modifying the stack image
-The `image` directory contains files for building the stack image. This image controls how the stack behaves throughout the application lifecycle.
+Stack creators configure the `APPSODY_DEPS` [environment variable](/content/docs/stacks/environment-variables.md) to specify the directory to be cached.
 
+## Application build and deployment
```
Comment (helenmasters) [inline content/docs/stacks/modify.md]: Some of these sections don't talk about modifying...?

File: content/docs/stacks/modify.md (hunk 1 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 2 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 3 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 4 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 5 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 6 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 8 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: lets elaborate on this in next pass.

File: content/docs/stacks/modify.md (hunk 9 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: there are a few here and I would like to properly describe them in the next pass

File: content/docs/stacks/modify.md (hunk 10 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: tried to address some of it.

File: content/docs/stacks/modify.md (hunk 11 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 12 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated

File: content/docs/stacks/modify.md (hunk 13 above)
Comment (neeraj-laad) [reply content/docs/stacks/modify.md]: updated
